contracts/TwabRewards.sol

```diff
@@ -0,0 +1,419 @@
+
+        if (_remainingRewards > 0) {
+            _token.safeTransfer(_to, _remainingRewards);
+        }
+
+        delete _promotions[_promotionId];
+
+        emit PromotionCancelled(_promotionId, _remainingRewards);
+
+        return true;
+    }
+
+    /// @inheritdoc ITwabRewards
+    function extendPromotion(uint256 _promotionId, uint8 _numberOfEpochs)
+        external
+        override
+        returns (bool)
+    {
+        require(_isPromotionActive(_promotionId) == true, "TwabRewards/promotion-not-active");
+
+        Promotion memory _promotion = _getPromotion(_promotionId);
+        uint8 _extendedNumberOfEpochs = _promotion.numberOfEpochs + _numberOfEpochs;
+
+        _requireEpochLimit(_extendedNumberOfEpochs);
+
+        IERC20 _token = IERC20(_promotion.token);
+        uint256 _amount = _numberOfEpochs * _promotion.tokensPerEpoch;
+
+        _token.safeTransfer(address(this), _amount);
+        _promotions[_promotionId].numberOfEpochs = _extendedNumberOfEpochs;
+
+        emit PromotionExtended(_promotionId, _amount, _extendedNumberOfEpochs);
+
+        return true;
+    }
+
+    /// @inheritdoc ITwabRewards
+    function getPromotion(uint256 _promotionId) external view override returns (Promotion memory) {
+        return _getPromotion(_promotionId);
+    }
+
+    /// @inheritdoc ITwabRewards
+    function getRemainingRewards(uint256 _promotionId) external view override returns (uint256) {
+        return _getRemainingRewards(_promotionId);
+    }
+
+    /// @inheritdoc ITwabRewards
+    function getCurrentEpochId(uint256 _promotionId) external view override returns (uint256) {
+        return _getCurrentEpochId(_promotionId);
+    }
+
+    /// @inheritdoc ITwabRewards
+    function getRewardAmount(
+        address _user,
+        uint256 _promotionId,
+        uint256[] calldata _epochIds
+    ) external view override returns (uint256[] memory) {
+        uint256[] memory _rewardsAmount = new uint256[](_epochIds.length);
+
+        for (uint256 index = 0; index < _epochIds.length; index++) {
+            _rewardsAmount[index] = _calculateRewardAmount(_user, _promotionId, _epochIds[index]);
+        }
+
+        return _rewardsAmount;
+    }
+
+    /// @inheritdoc ITwabRewards
+    function claimRewards(
+        address _user,
+        uint256 _promotionId,
+        uint256[] calldata _epochIds
+    ) external override returns (uint256) {
+        uint256 _rewardAmount;
+
+        for (uint256 index = 0; index < _epochIds.length; index++) {
+            uint256 _epochId = _epochIds[index];
+
+            require(_epochId < _getCurrentEpochId(_promotionId), "TwabRewards/epoch-not-over");
+
+            require(
+                !_isClaimedEpoch(_user, _promotionId, _epochId),
+                "TwabRewards/rewards-already-claimed"
+            );
+
+            _rewardAmount += _calculateRewardAmount(_user, _promotionId, _epochId);
+            _setClaimedEpoch(_claimedEpochs[_promotionId][_user], _epochId, true);
```

Comment:
how is this setting a flag when `_setClaimedEpoch()` is pure and the return value is not being stored?